### PR TITLE
Expose raw saved tensors for custom functions

### PR DESCRIPTION
**Pull request by Varal7**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60975
* #60663
* #60565
* #60685
* **#60551**

Summary: This PR is part of a stack aimed to address #58512. 
Provide bindings for the `SavedVariable` class (for custom functions only). The `SavedTensor` object can not be created from Python and currently does not expose any method.
Undefined behavior can arise if the user decides to keep a reference to this `SavedTensor` after the underlying `SavedVariable` has been freed.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466228](https://our.internmc.facebook.com/intern/diff/D29466228)

### CODE REVIEW

File: torch/csrc/autograd/python_function.cpp [hunk 1]
```diff
@@ -743,6 +743,25 @@ PyObject *THPFunction_saved_variables(THPFunction *self, void *_unused)
...
+  HANDLE_TH_ERRORS
+  auto& saved_variables = self->saved_variables;
+  if (saved_variables.empty())
+    return PyTuple_New(0);
+  int num_saved = saved_variables.size();
+  THPObjectPtr saved(PyTuple_New(num_saved));
+  if (!saved)
+    return nullptr;
+  for(const auto i : c10::irange(num_saved)) {
+    py::object obj = py::cast(saved_variables[i], py::return_value_policy::reference);
```
Comment (Varal7) [inline torch/csrc/autograd/python_function.cpp:757]: This should probably be abstracted away into another function

File: torch/csrc/autograd/python_function.cpp [hunk 2]
```diff
@@ -743,6 +743,25 @@ PyObject *THPFunction_saved_variables(THPFunction *self, void *_unused)
   END_HANDLE_TH_ERRORS
 }
 
+PyObject *THPFunction_raw_saved_tensors(THPFunction *self, void *_unused)
+{
+  HANDLE_TH_ERRORS
```
Comment (Varal7) [inline torch/csrc/autograd/python_function.cpp:748]: Maybe refactor this and unpack_saved_variables?

File: torch/csrc/autograd/python_function.cpp [hunk 3]
```diff
@@ -743,6 +743,25 @@ PyObject *THPFunction_saved_variables(THPFunction *self, void *_unused)
...
+    return nullptr;
+  for(const auto i : c10::irange(num_saved)) {
+    py::object obj = py::cast(saved_variables[i], py::return_value_policy::reference);
+    PyTuple_SET_ITEM(saved.get(), i, obj.release().ptr());
+  }
+  return saved.release();
+  END_HANDLE_TH_ERRORS
+}
+
+
```
Comment (Varal7) [inline torch/csrc/autograd/python_function.cpp:764]: Extra line

File: torch/csrc/autograd/python_function.cpp (hunk 1 above)
Comment (Varal7) [reply torch/csrc/autograd/python_function.cpp:757]: @soulitzer This snippet code is used three times but not sure where the best place to put it would be

Comment (soulitzer) [reply torch/csrc/autograd/python_function.cpp:757]: Where else is it being used?

Comment (soulitzer) [reply torch/csrc/autograd/python_function.cpp:757]: From here https://pybind11.readthedocs.io/en/stable/advanced/functions.html using py::return_value_policy::reference might lead to undefined behavior if the c++ side is freed and we keep using it on the Python side

File: torch/csrc/autograd/python_function.cpp [hunk 4]
```diff
@@ -845,6 +864,7 @@ PyObject* getRequiresGrad(PyObject* obj, void* _unused) {
 static struct PyGetSetDef THPFunction_properties[] = {
   {"saved_tensors", (getter)THPFunction_saved_tensors, nullptr, nullptr, nullptr},
   {"saved_variables", (getter)THPFunction_saved_variables, nullptr, nullptr, nullptr},
+  {"raw_saved_tensors", (getter)THPFunction_raw_saved_tensors, nullptr, nullptr, nullptr},
```
Comment (soulitzer) [inline torch/csrc/autograd/python_function.cpp:867]: Maybe add a small test to make sure we can access this from `grad_fn.raw_saved_tensor` and test what happens if there aren't any saved etc.

**Review by soulitzer — COMMENTED** [review]
Looks ok, the main issue I think is how to deal with the case when backward already frees the SavedVariable. Maybe theres some way to do this in pybind, but if we use the plain old PyObject struct, we can have it hold a weak_ptr and be able to raise an error if c++ side is already freed

**Comment by Varal7** [comment]
> Looks ok, the main issue I think is how to deal with the case when backward already frees the SavedVariable. Maybe theres some way to do this in pybind, but if we use the plain old PyObject struct, we can have it hold a weak_ptr and be able to raise an error if c++ side is already freed

Thanks for the review! 
Yes the behavior after #60565 is that accessing the SavedVariable will raise an error (https://github.com/pytorch/pytorch/pull/60565/files#diff-1fdfd061a92957fd11edbaceb538f7186f185b9fb22b7219107b007f7df557fbR750)
The error could be a tiny bit more explicit (in this case we know the user is explicitly trying to access the saved variable, and not trying to backward, but the current error message already mentions that it could be they are trying to access the saved variable).

I can rebase to include it in this PR.

**Comment by soulitzer** [comment]
Np!
Ahh yes, that error is probably good to include as well in this PR for completeness.
The situation I was thinking about is slightly different, i.e., if the user calls raw_saved_tensors and saves a reference to a SavedVariable, then calls backward.

**Comment by albanD** [comment]
Removing myself as @soulitzer is handling this one.
You can add me back if you have a specific question for me!

File: torch/csrc/autograd/python_function.cpp (hunk 2 above)
Comment (Varal7) [reply torch/csrc/autograd/python_function.cpp:748]: The function I meant to refactor was ihttps://github.com/pytorch/pytorch/blob/7e619b95880cdec72d1353c9c99dedbc2575ad36/torch/csrc/autograd/python_function.cpp#L689

(they already slightly diverged, but I don't have strong preference in favor of refactoring)

File: torch/csrc/autograd/python_function.cpp (hunk 1 above)
Comment (Varal7) [reply torch/csrc/autograd/python_function.cpp:759]: They are used [here](https://github.com/pytorch/pytorch/pull/60565/files#diff-2bf4105c8354f4a2656b421e0dc44d72542efdde41070ee67a81c7d99c34939aR216) and [here](https://github.com/pytorch/pytorch/pull/60565/files#diff-2bf4105c8354f4a2656b421e0dc44d72542efdde41070ee67a81c7d99c34939aR201)

**Comment by Varal7** [comment]
> Np!
> Ahh yes, that error is probably good to include as well in this PR for completeness.
> The situation I was thinking about is slightly different, i.e., if the user calls raw_saved_tensors and saves a reference to a SavedVariable, then calls backward.

Good point. My proposal is the following.
Because we only (and only plan to) expose the method `SavedVariable::register_hooks`, we can check there that the variable has not already been freed. 
So if the user keeps a reference to the raw_saved_tensors after he has called backward he will just hold a useless object and will have a nice error message if he tries to call register_hook on it.
Like so: https://github.com/pytorch/pytorch/pull/60663/files#diff-cc9fba479b5beae06b2eea2e390d17796e0341c5b037a20b5bcaccbb0c341030R156

**Comment by soulitzer** [comment]
That's a good plan - but there are also some cases where the entire SavedVariable is destroyed no (not just its data field)? 
For example all custom functions (we could add a test for this here), and some codegen'd functions store saved variable in a vector, and on `release_variables`, the vector is cleared.

**Comment by Varal7** [comment]
While I do agree with you, I've written tests to test those cases ([here](https://github.com/pytorch/pytorch/pull/60685/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR4801) and [here](https://github.com/pytorch/pytorch/pull/60565/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR4686)).

What is confusing to me is that the following code does trigger `RuntimeError: Calling register_hook on a saved tensor after it has been freed.` as we would want.

```python
class MyFn(Function):
    @staticmethod
    def forward(ctx, x):
        ctx.save_for_backward(x)
        return x

    @staticmethod
    def backward(ctx, g):
        return g, None

y = MyFn.apply(torch.ones(5, requires_grad=True))
t = y.grad_fn.raw_saved_tensors
y.sum().backward() # calls grad_fn->release_variables() and the destructors for each SavedVariable (but not `.reset()`)
t[0].register_hooks(lambda x: x, lambda x: x)

**Comment by soulitzer** [comment]
It could be because vector.clear just updates its size, but capacity is required to stays the same, and thus the memory is still reserved by the vector.
Maybe we should tell the vector `shrink_to_fit`, to actually release the memory (though technically that is still not guaranteed).
Could be interesting as well to see if this improves memory usage too

**Review by soulitzer — APPROVED** [review]
LGTM

File: torch/csrc/autograd/init.cpp [hunk 5]
```diff
@@ -263,6 +264,11 @@ PyObject* THPAutograd_initExtension(PyObject* _unused, PyObject *unused) {
   py::class_<DisableTorchDispatch>(_C_m, "_DisableTorchDispatch")
       .def(py::init<>());
 
+  py::class_<torch::autograd::SavedVariable>(m, "SavedTensor")
+    .def(py::init([]()->torch::autograd::SavedVariable {
+      TORCH_CHECK(false, "Trying to create a SavedTensor object from Python is forbidden.");
```
Comment (soulitzer) [inline torch/csrc/autograd/init.cpp:269]: Also add a test for this?

File: test/test_autograd.py [hunk 6]
```diff
@@ -4758,6 +4758,44 @@ def test_grad_fn_attr_bindings(self):
...
+            class MyFn(Function):
+                @staticmethod
+                def forward(ctx, x):
+                    if save_for_backward:
+                        ctx.save_for_backward(*eval(save_for_backward))
+                    return x
+
+                @staticmethod
+                def backward(ctx, g):
+                    return g, None
```
Comment (albanD) [inline test/test_autograd.py:4772]: Why is there two returns here while you have a single input for the forward?

Note to self: I still don't understand why we allow trailing None here :p

File: torch/csrc/autograd/python_function.cpp (hunk 4 above)
Comment (albanD) [reply torch/csrc/autograd/python_function.cpp:868]: Why is this public API? Shouldn't it be `_raw_saved_tensors` ?

**Review by albanD — COMMENTED** [review]
Just a couple of small comments.

File: torch/csrc/autograd/python_function.cpp (hunk 4 above)
Comment (Varal7) [reply torch/csrc/autograd/python_function.cpp:868]: I used `raw_saved_tensors` for consistency with the implicit rule "prefix `saved` with `raw_` and you get the raw version". Would you like me to change it?

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466228).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466228).

**Pull request closed without merging**